'static void IborCoupon::createIndexedCoupons();' instead of QL_USE_INDEXED_COUPON

ralfkonrad opened the pull request:
Hi all,

I see some discussions around the preprocessor flag `QL_USE_INDEXED_COUPON`.

One of the main disadvantages is the compile-time nature of this flag. In my opinion the results of calculation should not depend on preprocessor but should be configurable at run-time.

I suggest to extend the `Settings` class and use `Settings::instance().useIndexedCoupon()` to switch programmatically between par and indexed coupons.

I adjusted all code lines which link to `QL_USE_INDEXED_COUPON` to use `Settings::instance().useIndexedCoupon()` instead.

What do you think of the changes?


Best regards,
Ralf

lballabio commented:
Ralf, thanks.  I agree it would be desirable; the only concern I have is that it might decrease performance.  I'll try and make some measurements.

ralfkonrad commented:
Hi Luigi,

I have written a small performance test myself to create and calculate 125,000 swaps or to recalculate one swap 125,00 times.

I see a decrease of performance of 2.4% (29.679 s to 30.399 s) when creating the swaps but no performance impact when recalculating (3.0290 s to 3.0040 s).


Regards,
Ralf
 

```c++
/* -*- mode: c++; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- */

/*!
 Copyright (C) 2008 Allen Kuo

 This file is part of QuantLib, a free-software/open-source library
 for financial quantitative analysts and developers - http://quantlib.org/

 QuantLib is free software: you can redistribute it and/or modify it
 under the terms of the QuantLib license.  You should have received a
 copy of the license along with this program; if not, please email
 <quantlib-dev@lists.sf.net>. The license is also available online at
 <http://quantlib.org/license.shtml>.

 This program is distributed in the hope that it will be useful, but WITHOUT
 ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 FOR A PARTICULAR PURPOSE.  See the license for more details.
 */

/* This example sets up a callable fixed rate bond with a Hull White pricing
   engine and compares to Bloomberg's Hull White price/yield calculations.
*/

#include <ql/qldefines.hpp>
#ifdef BOOST_MSVC
#include <ql/auto_link.hpp>
#endif

#include <ql/indexes/ibor/euribor.hpp>
#include <ql/indexes/iborindex.hpp>
#include <ql/instruments/vanillaswap.hpp>
#include <ql/pricingengines/swap/discountingswapengine.hpp>
#include <ql/quotes/simplequote.hpp>
#include <ql/termstructures/yield/flatforward.hpp>
#include <ql/termstructures/yieldtermstructure.hpp>
#include <ql/time/daycounters/thirty360.hpp>
#include <boost/timer.hpp>
#include <iomanip>
#include <iostream>
#include <vector>

// This makes it easier to use array literals (alas, no std::vector literals)
#define LENGTH(a) (sizeof(a) / sizeof(a[0]))

using namespace QuantLib;

#if defined(QL_ENABLE_SESSIONS)
namespace QuantLib {
    Integer sessionId() { return 0; }
}
#endif

namespace {

    ext::shared_ptr<YieldTermStructure>
    flatRate(const Date& today, const ext::shared_ptr<Quote>& forward, const DayCounter& dc) {
        return ext::shared_ptr<YieldTermStructure>(
            new FlatForward(today, Handle<Quote>(forward), dc));
    }

    ext::shared_ptr<YieldTermStructure>
    flatRate(const Date& today, Rate forward, const DayCounter& dc) {
        return flatRate(today, ext::shared_ptr<Quote>(new SimpleQuote(forward)), dc);
    }

    struct CommonVars {
        // global data
        Date today, settlement;
        VanillaSwap::Type type;
        Real nominal;
        Calendar calendar;
        BusinessDayConvention fixedConvention, floatingConvention;
        Frequency fixedFrequency, floatingFrequency;
        DayCounter fixedDayCount;
        ext::shared_ptr<IborIndex> index;
        Natural settlementDays;
        RelinkableHandle<YieldTermStructure> termStructure;

        // cleanup
        SavedSettings backup;

        // utilities
        ext::shared_ptr<VanillaSwap>
        makeSwap(Integer length, Rate fixedRate, Spread floatingSpread) {
            Date maturity = calendar.advance(settlement, length, Years, floatingConvention);
            Schedule fixedSchedule(settlement, maturity, Period(fixedFrequency), calendar,
                                   fixedConvention, fixedConvention, DateGeneration::Forward,
                                   false);
            Schedule floatSchedule(settlement, maturity, Period(floatingFrequency), calendar,
                                   floatingConvention, floatingConvention, DateGeneration::Forward,
                                   false);
            ext::shared_ptr<VanillaSwap> swap(
                new VanillaSwap(type, nominal, fixedSchedule, fixedRate, fixedDayCount,
                                floatSchedule, index, floatingSpread, index->dayCounter()));
            swap->setPricingEngine(
                ext::shared_ptr<PricingEngine>(new DiscountingSwapEngine(termStructure)));
            return swap;
        }

        CommonVars() {
            type = VanillaSwap::Payer;
            settlementDays = 2;
            nominal = 100.0;
            fixedConvention = Unadjusted;
            floatingConvention = ModifiedFollowing;
            fixedFrequency = Annual;
            floatingFrequency = Semiannual;
            fixedDayCount = Thirty360();
            index =
                ext::shared_ptr<IborIndex>(new Euribor(Period(floatingFrequency), termStructure));
            calendar = index->fixingCalendar();
            today = calendar.adjust(Settings::instance().evaluationDate());
            settlement = calendar.advance(today, settlementDays, Days);
            termStructure.linkTo(flatRate(settlement, 0.01, Actual365Fixed()));
        }
    };
}


int main(int, char*[]) {
    try {

        boost::timer timer;

        Date today = Date(25, August, 2019);
        Settings::instance().evaluationDate() = today;

        CommonVars vars;

        Size tries = 1000;
        Integer lengths[] = {1, 2, 5, 10, 20};
        Rate rates[] = {0.0, 1.0, 2.0, 3.0, 4.0};
        Spread spreads[] = {-0.001, -0.01, 0.0, 0.01, 0.001};

        Size total = tries * LENGTH(lengths) * LENGTH(rates) * LENGTH(spreads);

        ext::shared_ptr<VanillaSwap> swap;
        for (Size i = 0; i < tries; i++) {
            for (Size j = 0; j < LENGTH(lengths); j++) {
                for (Size k = 0; k < LENGTH(rates); k++) {
                    for (Size l = 0; l < LENGTH(spreads); l++) {
                        ext::shared_ptr<VanillaSwap> swap =
                            vars.makeSwap(lengths[j], rates[k], spreads[l]);
                        swap->NPV();
                    }
                }
            }
        }

        double seconds = timer.elapsed();
        Integer hours = int(seconds / 3600);
        seconds -= hours * 3600;
        Integer minutes = int(seconds / 60);
        seconds -= minutes * 60;
        std::cout << total << " new created swaps completed in ";
        if (hours > 0)
            std::cout << hours << " h ";
        if (hours > 0 || minutes > 0)
            std::cout << minutes << " m ";
        std::cout << std::fixed << std::setprecision(4) << seconds << " s" << std::endl;

        timer.restart();

        swap = vars.makeSwap(5, 0.025, 0.001);
        for (Size i = 0; i < total; i++) {
            swap->NPV();
            swap->recalculate();
        }

        seconds = timer.elapsed();
        hours = int(seconds / 3600);
        seconds -= hours * 3600;
        minutes = int(seconds / 60);
        seconds -= minutes * 60;
        std::cout << "Recalculated swap " << total << " times completed in ";
        if (hours > 0)
            std::cout << hours << " h ";
        if (hours > 0 || minutes > 0)
            std::cout << minutes << " m ";
        std::cout << std::fixed << std::setprecision(4) << seconds << " s\n" << std::endl;

        return 0;

    } catch (std::exception& e) {
        std::cerr << e.what() << std::endl;
        return 1;
    } catch (...) {
        std::cerr << "unknown error" << std::endl;
        return 1;
    }
}
```

lballabio commented:
Thanks.

lballabio commented:
One thing I just realized: in the current implementation, if we write:
```
Settings::instance().useIndexedCoupon() = false;
IborCoupon coupon(...);
Settings::instance().useIndexedCoupon() = true;
```
the coupon will remain at par, even after setting the flag to true, because the flag is only checked in its constructor.  The same happens when we build, say, a floating-rate bond.

Now, I'm not sure what to think here.  The above might as well be the effect we want (that is, once built, the coupon stays the way it is) but then we should at least document it.  And we might also add the flag as an optional constructor parameter, since that's where it's used.  The settings would give it a default.

However, if we want the coupon to change instead (which could be argued) we need a way to update it.  I wouldn't move the check (and the date calculation that follows) to the calculation of the rate, because that would slow it down.  I'd probably use some kind of observer/observable to detect the change and recalculate the dates when it happens.  It's not straightforward, though.

All in all, I'd go for the first option.  What do you think?

ralfkonrad commented:
I agree, that's a valid point.

What about renaming `Settings::instance().useIndexedCoupon()` to `Settings::instance().createIndexedCoupons()` and have a `bool IborCoupon::isIndexedCoupon() const;` method:
```cpp
Settings::instance().createIndexedCoupons() = false;
IborCoupon coupon(...);
Settings::instance().createIndexedCoupons() = true;
...
if (coupon.isIndexedCoupon()) {
    ...
} else {
    ...
}
```
This makes it clearer, that only the constructor is affected and the behavior cannot be changed afterwards anymore.

I am not sure about the constructor flag in the `IborCoupon`. This suggests that we can have legs with mixed behaviours. Is that what we want?!

Only having `Settings::instance().createIndexedCoupons()` indicates the more "global" approach. I suggested this PR because we forgot to change the compiler flag before go live. This enables us to override the behavior via configuration at run time.

pcaspers commented:
If I may add a comment from my personal experience: I don't think the issue with QL_USE_INDEXED_COUPON is so much that a rebuild is necessary to switch back and forth, but rather that changing the flag triggers a *full* rebuild, although actually only a handful of files would actually require  a recompile. Not sure this is easy to solve though.

Switching the flag at runtime sounds like something I'd never dare to do except in small toy applications linking against QL, since it might cause all types of inconsistencies that will be hard to diagnose.

ralfkonrad commented:
> If I may add a comment from my personal experience:

Sure! ;-)

> I don't think the issue with QL_USE_INDEXED_COUPON is so much that a rebuild is necessary to switch back and forth, but rather that changing the flag triggers a _full_ rebuild, although actually only a handful of files would actually require a recompile.

The _full_ rebuild is one problem. But as we do not have a build pipeline with CI the time factor is not our biggest problem.

The other one was that we did not notice the possibility to switch the behavior right away. If the compiler flag is not set, the part of the code is grayed out and can be easily overlooked.

Also reference lookup does not work for preprocessors, at least in Visual Studio.

In other words: We did not expect that preprocessor flags control the result of a calcution.

All the other flags in userconfig.hpp are technical onces like QL_ERROR_FUNCTIONS, QL_ENABLE_SESSIONS or QL_EXTRA_SAFETY_CHECKS.

> Not sure this is easy to solve though.

_I admire the problem, but I do not have a solution._ :-)

> Switching the flag at runtime sounds like something I'd never dare to do except in small toy applications linking against QL,...

We do not want to switch the flag during calculation, but at startup which for me is part of runtime configuration. That is what we use the other settings for as well.

> ...since it might cause all types of inconsistencies that will be hard to diagnose.

But what is the difference to the other settings? If you change the `evaluationDate()` but have used the "wrong" term structure with fixed reference date, you also get _all types of inconsistencies_ as well. If we want to change the `evaluationDate()` we start a new process in our rick engine.

lballabio commented:
I'd also probably only use this for configuration at startup (and in other cases, trust that the users know what they're doing).  And yes, not triggering a full recompile would be nice, but a runtime flag would also make the choice available to those that don't compile the library at all (as in `pip install QuantLib-Python` or `apt-get install quantlib`).

This said, I'm still not sure of the details.  This is only used in `IborCoupon`, so ideally I'd like to put it there instead of in the general `Settings`. However, the latter does already have a few similar flags, so it's probably better if we do the non-ideal but consistent thing.

(Having a constructor argument was a tentative step towards restricting the scope to `IborCoupon`. I didn't envision a leg with different kind of coupons; I would have had the floating leg and/or bond also take the parameter and pass it down. But that's probably not worth it.)

In short: this PR does look like a good compromise.  But I'd document the thing more thoroughly.

ralfkonrad commented:
> This said, I'm still not sure of the details. This is only used in `IborCoupon`, so ideally I'd like to put it there instead of in the general `Settings`.

Luigi, you are talking about something like:
```cpp
    IborCoupon::Settings::instance().useIndexedCoupon() = true;
```
This could throw an exception if you change it after the first creation of an `IborCoupon` e.g. the first call of a constructor.

Should I prepare a new CR with this approach or change this one to `Settings::instance().createIndexedCoupons()` and `bool IborCoupon::isIndexedCoupon() const;` and "document the thing more thoroughly"?

ralfkonrad commented:
P.S.: I prefer
```cpp
    IborCoupon::Settings::instance().useIndexedCoupon() = true;
```
which is more like a pattern.

The _exception if you change it after the first creation_ would be optional.

lballabio commented:
I'd update this PR.  Let's keep the flag into `Settings`, together with the other ones already there. Renaming it to `createIndexedCoupons` is ok.  I'm not sure if we need `IborCoupon::isIndexedCoupon` for now, but in case, I'd probably invert it and call it `isAtPar` which is a wording that people actually use.

ralfkonrad commented:
Fine with me!

I try to update the PR this weekend.

pcaspers commented:
Do we see the performance drop negligible? Also asking because it was only tested for one single case.

ralfkonrad commented:
I cannot remember the exact times for the testsuite. But I think it was just a couple of seconds difference.

What would be other good test cases than?

pcaspers commented:
Other users running their stuff. Maybe make this change optional for the time being with a compiler flag so that people can try it out. Don't pay for what you don't use right? A few percent is not nothing.

ralfkonrad commented:
A compiler flag to make another compiler flag configurable?!

pcaspers commented:
yes.

ralfkonrad commented:
But to be serious: You are right that
> Don't pay for what you don't use right? A few percent is not nothing.

We are not using the QL as a high performance pricing engine rather than as a reference implementation on a high abstraction level. And from that perspective I suggested this PR.

lballabio commented:
Times (and timing code) are in [this earlier comment](https://github.com/lballabio/QuantLib/pull/665#issuecomment-524662406).  From the look of it, the loss should be amortized if one builds a swap and reprices it a few times.  @pcaspers, maybe you can get a chance to try this change in one of your XVA use cases?

pcaspers commented:
Sure I can do that, I'll let you know the results.

However we also have multithreading applications in multi-session mode. I'd expect that the call to a singleton in the Ibor coupon is more expensive there. I can not easily test their behaviour, that is a small project.

ralfkonrad commented:
I run the testsuite again and I do not see any difference in times:
```
compiler flag:
1>Tests completed in 11 m 37 s -> QL_USE_INDEXED_COUPON  undefined
1>Tests completed in 11 m 47 s -> QL_USE_INDEXED_COUPON    defined

setting:
1>Tests completed in 11 m 33 s -> QL_USE_INDEXED_COUPON  undefined
1>Tests completed in 11 m 35 s -> QL_USE_INDEXED_COUPON  undefined
1>Tests completed in 11 m 35 s -> QL_USE_INDEXED_COUPON    defined
1>Tests completed in 11 m 38 s -> QL_USE_INDEXED_COUPON    defined
```
I guess the differences depend more on the things happening in the background of my windows maschine.

pcaspers commented:
Yes, these timings do not say anything.

I did some tests on my side and don't see an impact on the timings of the examples I ran. I didn't expect any though since very few ibor coupons are created in comparison to the rest of the computations done. 

This doesn't mean there are other use cases. As I said I'd like to see a test where a lot of swaps a created in multiple sessions. I don't have the time to set this up though, unfortunately.

ralfkonrad commented:
> As I said I'd like to see a test where a lot of swaps a created in multiple sessions.

@pcaspers: Sounds interesting, I have used multithreading in c#, but not in c++. If you can give me a hint (e.g. a git repo) how you use multithreading and sessions I can try to set up such a test.

pcaspers commented:
sorry I have no public code to share regarding this use case; my basic concern is that here 

https://github.com/lballabio/QuantLib/blob/master/ql/patterns/singleton.hpp#L173

the singleton is looked up in a map and if this is big enough the cost for this might be substantial compared to what is happening in the Ibor coupon construction

this is just a theoretical thought, but I think it should be taken into consideration?

ralfkonrad commented:
> the singleton is looked up in a map and if this is big enough the cost for this might be substantial compared to what is happening in the Ibor coupon construction

But if the map lookup becomes a problem, my biggest concern would be `Settings::instance().evaluationDate()` as it really used everywhere!?

How many sessions do you use than?

I was expecting to have on session per thread with something like
```c++
Integer sessionId() {
    return std::hash<std::thread::id>{}(std::this_thread::
}
```

Running more threads at the same time as CPUS are available does not improve the performance anyhow, does it? So perhaps we need something like
```c++
static void Settting::ReleaseSession(Integer sessionId);
```
being called at the end of the thread to get rid of the session in the std::map?

pcaspers commented:
I am not aware of any places in the code with superfluous calls to `Settings::instance().evaluationDate()`, if there are any I'd agree that they should be deleted. 

I also don't know if the number of threads we are using is representative for the community, so I wouldn't make assumptions about that.

Look, all I am saying is that you already saw a performance hit (you mention 2.4% above for Ibor coupon creation) and that this might even be bigger in multi-session mode.

ralfkonrad commented:
Neither do I see any _superfluous calls_ to `Settings::instance().createIndexedCoupons()`.

Regarding the _2.4% performance hit_ we should take into that

- we are talking about ~0.00038ms on a 3GHz maschine per coupon,

- the test case is using a `FlatForward` termstructure, so there is no interpolation done compared to real world examples using a `ZeroCurve` etc and

- in production we would load the static data for each swap from a data store and write back the NPV to a data store.

So the real impact will be much smaller.

I think we need to decide if the _Setting itself_ is a step forward compared to the preprocessor flag taking into account the performance impact.

As you mentioned:

> Other users running their stuff.

We would definitely benefit from the flexibility having the option to switch between par and indexed coupons.

lballabio commented:
Ok, here's a proposal.

The goal is to make it possible to choose between the two implementations at run time instead of compile time (so for instance one can `pip install` a prebuilt Python wheel instead of recompiling it).

The use case should be as simple as possible: choose at the beginning of your program and forget it.  We document what happens if one flips the flag in mid-program, but we discourage it.  No support for different settings in different threads.  It basically works like the compile-time choice, except you can do it in a precompiled binary.

This way, we can keep it bare bones.  I'm thinking of a simple static bool variable in the `IborCoupon` class with a static getter and setter.  No singleton, no `instance()` method, no underlying map.  The compiler can inline the access away.

Thoughts?

pcaspers commented:
Sounds better to me anyway.

ralfkonrad commented:
Guess what: that's not my preferred solution! ;-)

But it is pragmatic.

What about this
```c++
public:
    static void createAtParCoupons() { usingAtParCoupons_ = true; }
    static void createIndexedCoupons()  { usingAtParCoupons_ = false; }
    static bool usingAtParCoupons() { return usingAtParCoupons_; }
private:
    static bool usingAtParCoupons_;
```
instead of a setter.

Do we than skip the un-static `bool isAtPar() const;` in the IborCoupon?

I will prepare a new PR for it the next days.

lballabio commented:
Yes, your interface is nicer.  Thanks.  If it's not too much of a bother, I'd modify this PR instead of creating a new one.  This way, we keep this discussion together with the changes.

pcaspers commented:
Just to explore the options, would an environment variable work as well?

ralfkonrad commented:
I guess an _environment variable_ will be possible but

- we loose the possibility to compare At Par with Indexed coupons in test scenarios directly without restarts and changing the variable.

- I guess we end up storing the value of the variable in `static bool usingAtParCoupons_ = ENV("USE_INDEXED_COUPON");`, so there will be no performance wins anyhow!?

- do you provide and test the implementation for all kind of platforms like *nix and Windows, Python/SWIG on *nix or Windows, Excel etc.?

pcaspers commented:
No I won't contribute anything to this PR except some wise comments ;-)

Would be kind of cool if a `static const bool` could be set from an environment variable? Never tried that, but it would prevent some of the problems discussed above. This solution could also coexist with the good old macro, i.e. the macro setting could be overridden by an env variable.

pcaspers commented:
Possibly (but I don't know) this also allows the compiler to essentially optimise away the case distinction?

pcaspers commented:
So what I meant with the possible optimisation: I thought here https://godbolt.org/z/x0rnw9 the compiler could generate two versions of func(), one that computes `a*2` and another that computes `a*3` and invoke the one corresponding to the value of `setting` that is (once and for all) determined from `getenv("setting")` at program start. This is apparently not the case. On the other hand this is a brittle example since removing `-fno-inline` completely optimises away the call to func() and the loop, so it's not valid to say that the compiler is missing a possible optimisation...

ralfkonrad commented:
@pcaspers I'm a little bit tired from arguing!

I am not happy with https://github.com/lballabio/QuantLib/pull/665#issuecomment-544883422
you are not happy with it,
so we have a compromise...

ralfkonrad commented:
@lballabio: I have done the changes, you can have a look.

lballabio commented:
Thanks, Ralf.  Before merging, I might try to have the tests with different results run both cases.  But the main code is ok for me.
